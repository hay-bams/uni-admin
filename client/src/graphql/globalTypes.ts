/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

export interface LoginInput {
  username?: string | null;
  password?: string | null;
  withCookie: boolean;
}

export interface NewCourseInput {
  name: string;
  totalSeats: number;
  status: string;
}

export interface NewStudentInput {
  email: string;
  country: string;
  name: string;
}

export interface RegisterInput {
  username: string;
  password: string;
}

export interface UpdateCourseInput {
  name: string;
  totalSeats: number;
  status: string;
}

//==============================================================
// END Enums and Input Objects
//==============================================================
