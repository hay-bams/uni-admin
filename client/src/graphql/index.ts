export * from './queries/Students'
export * from './queries/StudentDetails'
export * from './queries/AllCourses'
export * from './mutations/RegisterCourse'
export * from './mutations/Login'
export * from './mutations/Logout'