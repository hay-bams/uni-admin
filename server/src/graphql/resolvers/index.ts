import  merge from 'lodash.merge'
import { StudentsResolver } from './students'


export const resolvers = merge(StudentsResolver)