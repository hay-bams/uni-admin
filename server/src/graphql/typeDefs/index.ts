import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  input CourseInput {
    courseId: ID!
  }

  input NewStudentInput {
    email: String!,
    country: String!, 
    name: String!
  }

  input NewCourseInput {
    name: String!
    totalSeats: Int!
    status: String!
  }

  input LoginInput {
    username: String
    password: String
    withCookie: Boolean!
  }

  input RegisterInput {
    username: String!
    password: String!
  }

  type Student {
    id: ID!
    studentID: String!
    status: String!
    name: String!
    email: String!
    country: String!
    courses: [Course!]!
  }

  type Students {
    total: Int
    results: [Student!]!
  }

  type Courses {
    total: Int
    results: [Course!]!
  }

  type Course {
    id: ID!
    name: String!
    totalSeats: Int!
    status: String!
  }

  type StudentData {
    student: Student!
    courses: [Course!]!
  }

  type User {
    id: ID
    username: String
    token: String 
    madeRequest: Boolean
  }

  type Query {
    students(all: String, limit: Int, page: Int): Students!
    studentDetails(id: ID!): Student!
    allCourses(all: String, limit: Int, page: Int): Courses!
  }

  type Mutation {
    registerCourse(studentId: String!, input: ID!): Student
    unregisterCourse(studentId: String!, courseId: String): Student
    login(input: LoginInput): User!
    logout: User!
    register(input: RegisterInput): User!
    addNewStudent(input: NewStudentInput): Student!
    addNewCourse(input: NewCourseInput): Course!
  }
`;
