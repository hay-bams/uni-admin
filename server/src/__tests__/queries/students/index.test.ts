/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { COURSES_DATA, STUDENT_DATA } from '../../test_data';
import { createTestServer } from '../../helpers';
import {
  TEST_ALL_STUDENTS_QUERY,
  TEST_STUDENT_DETAILS_QUERY,
} from '../queries';
import { GraphQLError } from 'graphql';

describe('Students Queries', () => {
  test('should return all students', async () => {
    const { query } = createTestServer({
      db: {
        students: {
          find: jest.fn(() => ({
            count: jest.fn(() => 4),
            toArray: jest.fn(() => STUDENT_DATA),
          })),
        },
      },
    });

    const res = await query({
      query: TEST_ALL_STUDENTS_QUERY,
      variables: { all: 'all' },
    });

    expect(res.data.students.total).toBe(STUDENT_DATA.length);
    expect(res.data.students.results.length).toBeGreaterThan(0);
    expect(res.data.students.results[0]).toHaveProperty('id');
    expect(res).toMatchSnapshot();
  });

  test('should return students with pagination', async () => {
    const { query } = createTestServer({
      db: {
        students: {
          find: jest.fn(() => ({
            skip: jest.fn(() => ({
              limit: jest.fn(() => ({
                count: jest.fn(() => 4),
                toArray: jest.fn(() => STUDENT_DATA),
              })),
            })),
          })),
        },
      },
    });

    const res = await query({
      query: TEST_ALL_STUDENTS_QUERY,
      variables: { page: 1, limit: 2 },
    });

    expect(res.data.students.total).toBe(STUDENT_DATA.length);
    expect(res.data.students.results.length).toBeGreaterThan(0);
    expect(res.data.students.results[0]).toHaveProperty('id');
    expect(res).toMatchSnapshot();
  });

  test('should return student details', async () => {
    const { query } = createTestServer({
      db: {
        students: {
          findOne: jest.fn(
            () => new Promise((resolve) => resolve(STUDENT_DATA[0]))
          ),
        },
        courses: {
          find: jest.fn(() => ({
            toArray: jest.fn(() => COURSES_DATA),
          })),
        },
      },
    });

    const res = await query({
      query: TEST_STUDENT_DETAILS_QUERY,
      variables: { id: STUDENT_DATA[0]._id.toString() },
    });

    expect(res.data.studentDetails).toHaveProperty('id');
    expect(res.data.studentDetails).toHaveProperty('name');
    expect(res.data.studentDetails.courses.length).toBeGreaterThan(0);
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if student is not found', async () => {
    const { query } = createTestServer({
      db: {
        students: {
          findOne: jest.fn(() => null),
        },
      },
    });

    const res = await query({
      query: TEST_STUDENT_DETAILS_QUERY,
      variables: { id: STUDENT_DATA[0]._id.toString() },
    });
    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if a graphql server error occurs while fetching students', async () => {
    const { query } = createTestServer({
      err: new GraphQLError('Some error occured'),
    });

    const res = await query({
      query: TEST_ALL_STUDENTS_QUERY,
      variables: { all: 'all' },
    });

    expect(res.errors!.length).toBeGreaterThan(0);
  });
});
