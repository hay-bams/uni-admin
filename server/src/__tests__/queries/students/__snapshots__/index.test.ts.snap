// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Students Queries should return all students 1`] = `
Object {
  "data": Object {
    "students": Object {
      "results": Array [
        Object {
          "id": "5fda72349baf026a892dc1a0",
        },
        Object {
          "id": "5fe36133a29a42f8111cf29c",
        },
        Object {
          "id": "5fe3613b0818b75505901627",
        },
        Object {
          "id": "5fe3617932e06a4c602772d0",
        },
      ],
      "total": 4,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Students Queries should return student details 1`] = `
Object {
  "data": Object {
    "studentDetails": Object {
      "courses": Array [
        Object {
          "id": "5fdab398791e134a5466566f",
        },
        Object {
          "id": "5fdab4b5eed1fb397feaca14",
        },
        Object {
          "id": "5fdab4bd78cc9321b31faf35",
        },
        Object {
          "id": "5fdab4c4fe1aa062abc46096",
        },
        Object {
          "id": "5fdab4ccba6dcf525408412a",
        },
      ],
      "id": "5fda72349baf026a892dc1a0",
      "name": "Peace Dike",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Students Queries should return students with pagination 1`] = `
Object {
  "data": Object {
    "students": Object {
      "results": Array [
        Object {
          "id": "5fda72349baf026a892dc1a0",
        },
        Object {
          "id": "5fe36133a29a42f8111cf29c",
        },
        Object {
          "id": "5fe3613b0818b75505901627",
        },
        Object {
          "id": "5fe3617932e06a4c602772d0",
        },
      ],
      "total": 4,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Students Queries should throw an error if student is not found 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Something went wrong: Error: Student not found],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
